MENTAL-MODEL.md

```diff
@@ -0,0 +1,376 @@
+```
+
+📤 Sending payjoin is just that simple.
+
+## Receive a Payjoin
+
+The `receiver` feature provides all of the check methods, PSBT data manipulation, coin selection, and transport structures to receive payjoin and handle errors in a privacy preserving way.
+
+Receiving payjoin entails listening to a secure http endpoint for inbound requests. The endpoint is displayed in the `pj` parameter of a bip21 request URI.
+
+The reference implementation uses `rouille` sync http server and Bitcoin Core RPC.
+
+```rs
+fn receive_payjoin(
+    bitcoind: bitcoincore_rpc::Client,
+    amount_arg: &str,
+    endpoint_arg: &str,
+) -> Result<()>
+```
+
+### 1. Generate a pj_uri BIP21 using `payjoin::Uri::from_str`
+
+A BIP 21 URI supporting payjoin contains at minimum a bitcoin address and a secure `pj` endpoint.
+
+```rs
+let pj_receiver_address = bitcoind.get_new_address(None, None)?;
+let amount = Amount::from_sat(amount_arg.parse()?);
+let pj_uri_string = format!(
+    "{}?amount={}&pj={}",
+    pj_receiver_address.to_qr_uri(),
+    amount.to_btc(),
+    endpoint_arg
+);
+let pj_uri = Uri::from_str(&pj_uri_string)
+    .map_err(|e| anyhow!("Constructed a bad URI string from args: {}", e))?;
+let _pj_uri = pj_uri
+    .check_pj_supported()
+    .map_err(|e| anyhow!("Constructed URI does not support payjoin: {}", e))?;
+```
+
+### 2. Listen for a sender's request on the `pj` endpoint
+
+Start a server to respond to payjoin protocol POST messages.
+
+```rs
+rouille::start_server("0.0.0.0:3000", move |req| handle_web_request(&req, &bitcoind));
+// ...
+fn handle_web_request(req: &Request, bitcoind: &bitcoincore_rpc::Client) -> Response {
+    handle_payjoin_request(req, bitcoind)
+        .map_err(|e| match e {
+            ReceiveError::RequestError(e) => {
+                log::error!("Error handling request: {}", e);
+                Response::text(e.to_string()).with_status_code(400)
+            }
+            e => {
+                log::error!("Error handling request: {}", e);
+                Response::text(e.to_string()).with_status_code(500)
+            }
+        })
+        .unwrap_or_else(|err_resp| err_resp)
+}
+```
+
+### 3. Parse an incoming request using `UncheckedProposal::from_request()`
+
+Parse incoming HTTP request and check that it follows protocol.
+
+```rs
+let headers = Headers(req.headers());
+let proposal = payjoin::receiver::UncheckedProposal::from_request(
+    req.data().context("Failed to read request body")?,
+    req.raw_query_string(),
+    headers,
+)?;
+```
+
+Headers are parsed using the `payjoin::receiver::Headers` Trait so that the library can iterate through them, ideally without cloning.
+
+```rs
+struct Headers<'a>(rouille::HeadersIter<'a>);
+impl payjoin::receiver::Headers for Headers<'_> {
+    fn get_header(&self, key: &str) -> Option<&str> {
+        let mut copy = self.0.clone(); // lol
+        copy.find(|(k, _)| k.eq_ignore_ascii_case(key)).map(|(_, v)| v)
+    }
+}
+```
```

Comment:
I reckon bindings will run into a problem with this Trait-based design. @thunderbiscuit, what do you think?